lib/services/sigstore_policies.go

```diff
@@ -155,14 +155,30 @@ func ValidateSigstorePolicy(s *workloadidentityv1pb.SigstorePolicy) error {
 			}
 		}
 
+		var roots root.TrustedMaterialCollection
 		for idx, trustedRoot := range v.Keyless.GetTrustedRoots() {
-			if _, err := root.NewTrustedRootFromJSON([]byte(trustedRoot)); err != nil {
+			root, err := root.NewTrustedRootFromJSON([]byte(trustedRoot))
+			if err != nil {
 				return trace.BadParameter("spec.keyless.trusted_roots[%d]: failed to parse trusted root: %v", idx, err)
 			}
+			roots = append(roots, root)
 		}
+		if len(roots) != 0 && len(roots.CTLogs()) == 0 && len(roots.TimestampingAuthorities()) == 0 {
+			return trace.BadParameter("spec.keyless.trusted_roots: must configure at least one transparency log or timestamp authority")
+		}
```

Comment:
It looks like having zero trusted roots is fine according to this check. Double checking this is correct?

Also, current implementation of `CTLogs()` works just fine with nil `tmc`, but I'm not sure how stable that property is.

```
func (tmc TrustedMaterialCollection) CTLogs() map[string]*TransparencyLog {
	rekorLogs := make(map[string]*TransparencyLog)
	for _, tm := range tmc {
		for keyID, tlogVerifier := range tm.CTLogs() {
			rekorLogs[keyID] = tlogVerifier
		}
	}
	return rekorLogs
}
```

```
func TestNilTMC(t *testing.T) {
	var roots root.TrustedMaterialCollection
	require.NotNil(t, roots.CTLogs())
}
```

Reply:
Yeah - if this logic is as it should be, then a short comment to explain the check may be helpful.

Reply:
Great call! I've added a short comment to explain this.

By default, we use the Sigstore project's "public good" [trusted roots](https://github.com/sigstore/root-signing) and this check validates any custom roots the user overrides them with.

Also good catch with the `nil` TMC thing. I've swapped the `var` declaration with a `make` assignment.

Reply:
Actually the test above passes 😄 I think we can keep it either nil or not, doesn't matter as we only look at contents when `roots` is not nil, and even then `CTLogs()` doesn't panic for nil `roots`. Up to you how you want it though.